File: leetcode/0134-GasStation/solution.go
// Space complexity O(N)
// Compute complexity O(N)

func canCompleteCircuit(gas []int, cost []int) int {
    extraGasTillStart := -1
    startIdx := 0
    sumExtraGas := 0

    for idx, g := range gas {
        currExtraGas := g - cost[idx]
        sumExtraGas += currExtraGas
        if extraGasTillStart < 0 {
            startIdx = idx
            extraGasTillStart = currExtraGas
        } else {
            extraGasTillStart += currExtraGas
        }
    }

    if sumExtraGas < 0 {
        return -1
    }

    return startIdx
}

File: leetcode/0134-GasStation/solution.cpp
// Compute Complexity O(N)
// Space Complexity O(N)

class Solution {
    public:
        int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
                   auto sumExtraGas = 0;
            int bestStart = 0;
    
            const auto size = gas.size();
            auto bestExtraGas = - 1;
    
            for (auto i = 0; i < size; i++) {
                const auto  currExtraGas = gas[i] - cost[i];
                sumExtraGas += currExtraGas;
    
                if (bestExtraGas < 0) {
                    bestStart = i;
                    bestExtraGas = currExtraGas;
                } else {
                    bestExtraGas += currExtraGas;
                }
            }
    
            if (sumExtraGas < 0) {
                return -1;
            }
    
            return bestStart;
        }
    };